### PR TITLE
Support more ops in post training quantization, test=develop

### PR DESCRIPTION
**Pull request by juncaipeng**
* Support  more ops in post training quantization, and save the output scale in quantized op.
* Update docs in post training quantization and qat

### CODE REVIEW

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 1]
```diff
@@ -1183,77 +1192,141 @@ def _scale_name(self, var_name):
...
+    supported_quantizable_op_type = [
+        "pool2d", "elementwise_add", "concat", "softmax", "argmax", "transpose",
+        "equal", "gather", "greater_equal", "greater_than", "less_equal",
+        "less_than", "mean", "not_equal", "reshape", "reshape2",
+        "bilinear_interp", "nearest_interp", "trilinear_interp", "slice",
+        "squeeze", "elementwise_sub"
+    ]
+    # list op real input and output names, to avoid processing input
+    # such as AxisTensor.
+    op_real_in_out_name = {
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: `op_real_in_out_name`变量外部不使用请加下划线。放在类外面，并改名为`_op_in_out_param_names`。

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py [hunk 2]
```diff
@@ -41,8 +41,9 @@ def __init__(self,
                  algo="KL",
                  quantizable_op_type=[
                      "conv2d", "depthwise_conv2d", "mul", "pool2d",
-                     "elementwise_add"
-                 ]):
+                     "elementwise_add", "concat"
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py]: `"pool2d","elementwise_add", "concat"`这三个op在lite上还没有相应int8kernel的支持，建议先不要将它们放在默认量化op表中。

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py [hunk 3]
```diff
@@ -66,7 +67,14 @@ def __init__(self,
                 abs_max methon to get the scale factor. Default is KL.
             quantizable_op_type(list[str], optional): List the type of ops 
                 that will be quantized. Default is ["conv2d", "depthwise_conv2d", 
-                "mul", "pool2d", "elementwise_add"].
+                "mul", "pool2d", "elementwise_add", "concat"].
+            is_full_quantized(bool, optional): If set is_full_quantized be True,
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py]: If set is_full_quantized as True,

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py [hunk 4]
```diff
@@ -66,7 +67,14 @@ def __init__(self,
                 abs_max methon to get the scale factor. Default is KL.
             quantizable_op_type(list[str], optional): List the type of ops 
                 that will be quantized. Default is ["conv2d", "depthwise_conv2d", 
-                "mul", "pool2d", "elementwise_add"].
+                "mul", "pool2d", "elementwise_add", "concat"].
+            is_full_quantized(bool, optional): If set is_full_quantized be True, 
+                apply quantization to all supported quantizable op type. If set 
+                is_full_quantized be False, only apply quantization to the op type
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py]: If set  is_full_quantized as False

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 5]
```diff
@@ -1183,77 +1192,141 @@ def _scale_name(self, var_name):
 
 
 class AddQuantDequantPass(object):
+    supported_quantizable_op_type = [
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: `supported_quantizable_op_type`变量外部不使用请加下划线。

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py [hunk 6]
```diff
@@ -340,6 +371,19 @@ def _set_var_node_value(self, var_node_name, np_value):
             tensor = var_node.get_tensor()
             tensor.set(np_value, self._place)
 
+    def _is_input_all_not_persistable(self, op, persistable_var_names):
+        '''
+        Analyse the real inputs of the op are all not persistable.
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py]: Analyze

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py [hunk 7]
```diff
@@ -269,38 +287,37 @@ def _update_program(self):
...
         for var in self._program.list_vars():
             if var.name in self._quantized_act_var_name:
                 var.persistable = False
 
         # use QuantizationTransformPass to insert fake_quantize/fake_dequantize op
         graph = IrGraph(core.Graph(self._program.desc), for_test=True)
 
-        qtp_quantizable_op_type = []
-        for op_type in ["conv2d", "depthwise_conv2d", "mul"]:
+        g_act_weight_quantizable_op_type = []
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py]: `g_act_weight_quantizable_op_type ` --> `major_quantizable_op_types `

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py [hunk 8]
```diff
@@ -269,38 +287,37 @@ def _update_program(self):
...
             activation_quantize_type='moving_average_abs_max',
             weight_quantize_type='channel_wise_abs_max',
-            quantizable_op_type=qtp_quantizable_op_type)
+            quantizable_op_type=g_act_weight_quantizable_op_type)
         transform_pass.apply(graph)
 
         # use AddQuantDequantPass to insert fake_quant_dequant op
-        aqdp_quantizable_op_type = []
-        for op_type in ["pool2d", "elementwise_add"]:
+        g_act_quantizable_op_type = []
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py]: `g_act_quantizable_op_type` --> `minor_quantizable_op_types`

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 9]
```diff
@@ -1183,77 +1192,141 @@ def _scale_name(self, var_name):
...
+        "equal", "gather", "greater_equal", "greater_than", "less_equal",
+        "less_than", "mean", "not_equal", "reshape", "reshape2",
+        "bilinear_interp", "nearest_interp", "trilinear_interp", "slice",
+        "squeeze", "elementwise_sub"
+    ]
+    # list op real input and output names, to avoid processing input
+    # such as AxisTensor.
+    op_real_in_out_name = {
+        "conv2d": [["Input"], ["Output"]],
+        "depthwise_conv2d": [["Input"], ["Output"]],
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: [["Input", "Filter"], ["Output"]],

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py [hunk 10]
```diff
@@ -319,10 +336,24 @@ def _update_program(self):
...
 
+    def _save_output_scale(self):
+        '''
+        Save output scale to the quantized op.
+        '''
+        output_scale_name = "output_scale"
+        for op in self._program.global_block().ops:
+            if op.type in self._quantizable_op_type:
+                output_name_list = self._op_real_in_out_name[op.type][1]
+                output_var_name = op.output(output_name_list[0])[0]
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py]: 改为：
```python
output_name_list = self._op_real_in_out_name[op.type][1]
for output_var_name in output_name_list:
```

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 11]
```diff
@@ -54,6 +54,8 @@ def _init_var_node(var_node, value, scope, place):
 
 
 class QuantizationTransformPass(object):
+    supported_quantizable_op_type = ['conv2d', 'depthwise_conv2d', 'mul']
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: 变量外部用户不使用请加下划线。

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 12]
```diff
@@ -589,32 +594,32 @@ def _quantized_scale_name(self, var_name):
 
 
 class QuantizationFreezePass(object):
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py:630]: 添加
`_supported_quantizable_op_type = QuantizationTransformPass._supported_quantizable_op_type`

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 13]
```diff
@@ -625,9 +630,8 @@ def __init__(self,
         self._activation_bits = activation_bits
         self._weight_quantize_type = weight_quantize_type
         self._quantizable_ops = quantizable_op_type
-        supported_quantizable_ops = ['conv2d', 'depthwise_conv2d', 'mul']
         for op in self._quantizable_ops:
-            assert op in supported_quantizable_ops, \
+            assert op in QuantizationTransformPass.supported_quantizable_op_type, \
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: `assert op in QuantizationFreezePass.supported_quantizable_op_type, \`

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 14]
```diff
@@ -903,7 +909,8 @@ class ConvertToInt8Pass(object):
         place(fluid.CPUPlace|fluid.CUDAPlace): place is used to restore the
             8bits weight tensors.
         quantizable_op_type(list[str]): List the type of ops that will be quantized. 
-            Default is ["conv2d", "depthwise_conv2d", "mul"].
+            Default is ["conv2d", "depthwise_conv2d", "mul"]. The quantizable_op_type in
+            QuantizationTransformPass and QuantizationFreezePass must be the same as this.
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: 添加
`_supported_quantizable_op_type = QuantizationTransformPass._supported_quantizable_op_type`

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 15]
```diff
@@ -917,9 +924,8 @@ def __init__(self,
         self._scope = scope
         self._place = place
         self._quantizable_ops = quantizable_op_type
-        supported_quantizable_ops = ['conv2d', 'depthwise_conv2d', 'mul']
         for op in self._quantizable_ops:
-            assert op in supported_quantizable_ops, \
+            assert op in QuantizationTransformPass.supported_quantizable_op_type, \
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: `assert op in ConvertToInt8Pass._supported_quantizable_op_type, \`

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py [hunk 16]
```diff
@@ -98,21 +106,26 @@ def __init__(self,
...
-        supported_quantizable_op_type = [
-            "conv2d", "depthwise_conv2d", "mul", "pool2d", "elementwise_add"
-        ]
-        for op_type in self._quantizable_op_type:
-            assert op_type in supported_quantizable_op_type, \
-                op_type + " is not supported for quantization."
+
+        supported_quantizable_op_type = \
+            QuantizationTransformPass.supported_quantizable_op_type + \
+            AddQuantDequantPass.supported_quantizable_op_type
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py]:
```python
        supported_quantizable_op_type = (
                 QuantizationTransformPass.supported_quantizable_op_type + 
                 AddQuantDequantPass.supported_quantizable_op_type)
```

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 17]
```diff
@@ -1183,77 +1192,141 @@ def _scale_name(self, var_name):
...
                  moving_rate=0.9,
                  quant_bits=8,
                  skip_pattern='skip_quant',
-                 quantizable_op_type=["elementwise_add", "pool2d"]):
+                 quantizable_op_type=["elementwise_add", "pool2d", "concat"],
+                 is_full_quantized=False):
         """
-        This pass is used to add quant_dequant op for some ops, such as the
-        'elementwise_add' and 'pool2d' op.
+        This pass add quant_dequant op for some ops, of which the inputs must be
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: of which all the inputs must be

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 18]
```diff
@@ -1183,77 +1192,141 @@ def _scale_name(self, var_name):
...
                  skip_pattern='skip_quant',
-                 quantizable_op_type=["elementwise_add", "pool2d"]):
+                 quantizable_op_type=["elementwise_add", "pool2d", "concat"],
+                 is_full_quantized=False):
         """
-        This pass is used to add quant_dequant op for some ops, such as the
-        'elementwise_add' and 'pool2d' op.
+        This pass add quant_dequant op for some ops, of which the inputs must be 
+        not persistable.
+        The quant_dequant op can obtain the input scale.
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: The input scales can be obtained from the quant_dequant op.

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 19]
```diff
@@ -1183,77 +1192,141 @@ def _scale_name(self, var_name):
...
             '%s_grad' % (op) for op in self._quantizable_op_type
         ]
 
-        supported_quantizable_op_type = ["elementwise_add", "pool2d"]
-        for op_type in quantizable_op_type:
-            assert op_type in supported_quantizable_op_type, \
-                op_type + " is not supported for quantization."
+        assert self._scope != None, "scope must not be None."
+        assert self._place != None, "place must not be None."
+        assert self._quant_bits == 8, "quant_bits must be 8 for now."
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: Don't need this 8 bit assert.

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 20]
```diff
@@ -1183,77 +1192,141 @@ def _scale_name(self, var_name):
...
         """
         assert isinstance(graph,
                           IrGraph), 'graph must be the instance of IrGraph.'
         self._is_test = graph.is_test()
         dequantized_vars_map = collections.OrderedDict()
-        ops = graph.all_op_nodes()
 
-        for op_node in ops:
+        # Forward stage, insert quant_dequant op
+        for op_node in graph.all_op_nodes():
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: 改回：
```python
ops = graph.all_op_nodes()
for op_node in ops:
```

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 21]
```diff
@@ -1266,6 +1339,22 @@ def apply(self, graph):
...
+        is_input_all_not_persistable = True
+        op_node_name = op_node.name()
+
+        input_name_list = \
+            AddQuantDequantPass.op_real_in_out_name[op_node_name][0]
+        for input_name in input_name_list:
+            for arg_name in op_node.input(input_name):
+                in_node = graph._find_node_by_name(op_node.inputs, arg_name)
+                is_input_all_not_persistable = is_input_all_not_persistable and \
+                    (not in_node.persistable())
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]:
```python
                is_input_all_not_persistable = (is_input_all_not_persistable and
                    (not in_node.persistable()))
```

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 22]
```diff
@@ -1183,77 +1192,141 @@ def _scale_name(self, var_name):
...
+                    op_node.name()][0]
+                for input_name in input_name_list:
+                    for arg_name in op_node.input(input_name):
+                        in_node = graph._find_node_by_name(op_node.inputs,
+                                                           arg_name)
+                        if arg_name in dequantized_vars_map:
+                            quant_var_node = dequantized_vars_map[arg_name]
+                        else:
+                            quant_var_node, _ = \
+                                self._inser_quant_dequant_moving_average_abs_max_op(
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py:1333]: `self._inser_quant_dequant_moving_average_abs_max_op`第二个返回值若无用，可不返回。

**Review by wzzju — CHANGES_REQUESTED** [review]

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 22 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py:1333]: > `self._inser_quant_dequant_moving_average_abs_max_op`第二个返回值若无用，可不返回。
不改原先的代码吧

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 1 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: > `op_real_in_out_name`变量外部不使用请加下划线。放在类外面，并改名为`_op_in_out_param_names`。

Done

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py (hunk 2 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py]: > `"pool2d","elementwise_add", "concat"`这三个op在lite上还没有相应int8kernel的支持，建议先不要将它们放在默认量化op表中。

Done

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py (hunk 3 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py]: > If set is_full_quantized as True,

Done

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py (hunk 4 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py]: > If set is_full_quantized as False

Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 5 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: > `supported_quantizable_op_type`变量外部不使用请加下划线。

Done

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py (hunk 6 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py]: > Analyze

两者都一样的意思

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py (hunk 7 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py]: > `g_act_weight_quantizable_op_type ` --> `major_quantizable_op_types `

Done

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py (hunk 8 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py]: > `g_act_quantizable_op_type` --> `minor_quantizable_op_types`

Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 9 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: > [["Input", "Filter"], ["Output"]],

Done

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py (hunk 10 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py]: > 改为：
> 
> ```python
> output_name_list = self._op_real_in_out_name[op.type][1]
> for output_var_name in output_name_list:
> ```

Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 11 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: > 变量外部用户不使用请加下划线。

Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 12 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py:630]: > 添加
> `_supported_quantizable_op_type = QuantizationTransformPass._supported_quantizable_op_type`

Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 13 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: > `assert op in QuantizationFreezePass.supported_quantizable_op_type, \`

Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 14 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: > 添加
> `_supported_quantizable_op_type = QuantizationTransformPass._supported_quantizable_op_type`

Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 15 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: > `assert op in ConvertToInt8Pass._supported_quantizable_op_type, \`

Done

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py (hunk 16 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py]: > ```python
>         supported_quantizable_op_type = (
>                  QuantizationTransformPass.supported_quantizable_op_type + 
>                  AddQuantDequantPass.supported_quantizable_op_type)
> ```

Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 17 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: > of which all the inputs must be

Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 18 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: > The input scales can be obtained from the quant_dequant op.

Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 19 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: > Don't need this 8 bit assert.

Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 20 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: > 改回：
> 
> ```python
> ops = graph.all_op_nodes()
> for op_node in ops:
> ```

Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 21 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: > ```python
>                 is_input_all_not_persistable = (is_input_all_not_persistable and
>                     (not in_node.persistable()))
> ```

一个意思吧？

**Comment by luotao1** [comment]
image
https://www.paddlepaddle.org.cn/documentation/docs/zh/advanced_usage/development/contribute_to_paddle/submit_pr_guide.html#id5 @juncaipeng

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 21 above)
Comment (wzzju) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: > > ```python
> >                 is_input_all_not_persistable = (is_input_all_not_persistable and
> >                     (not in_node.persistable()))
> > ```
> 
> 一个意思吧？

请查阅谷歌编码规范。

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 12 above)
Comment (wzzju) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py:630]: Not done. 添加QuantizationFreezePass的类变量`_supported_quantizable_op_type = QuantizationTransformPass._supported_quantizable_op_type`。

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 23]
```diff
@@ -625,9 +659,8 @@ def __init__(self,
         self._activation_bits = activation_bits
         self._weight_quantize_type = weight_quantize_type
         self._quantizable_ops = quantizable_op_type
-        supported_quantizable_ops = ['conv2d', 'depthwise_conv2d', 'mul']
         for op in self._quantizable_ops:
-            assert op in supported_quantizable_ops, \
+            assert op in QuantizationTransformPass._supported_quantizable_op_type, \
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: `assert op in QuantizationFreezePass._supported_quantizable_op_type, \`

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 14 above)
Comment (wzzju) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: Not done. 添加ConvertToInt8Pass的类变量`_supported_quantizable_op_type = QuantizationTransformPass._supported_quantizable_op_type`。

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 24]
```diff
@@ -917,9 +953,8 @@ def __init__(self,
         self._scope = scope
         self._place = place
         self._quantizable_ops = quantizable_op_type
-        supported_quantizable_ops = ['conv2d', 'depthwise_conv2d', 'mul']
         for op in self._quantizable_ops:
-            assert op in supported_quantizable_ops, \
+            assert op in QuantizationTransformPass._supported_quantizable_op_type, \
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: `assert op in ConvertToInt8Pass._supported_quantizable_op_type, \`

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 25]
```diff
@@ -1183,77 +1221,111 @@ def _scale_name(self, var_name):
...
+            moving_rate(float, optional): the param for 'quant_dequant_moving_average_abs_max' 
+                quantization. Default is 0.9.
+            quant_bits(int, optional): quantization bit number for activation. Default is 8.
+            skip_pattern(str, optional): The user-defined quantization skip pattern, which
+                will be presented in the name scope of an op. When the skip pattern is
+                detected in an op's name scope, the corresponding op will not be quantized.
+                Default is 'skip_quant'.
+            quantizable_op_type(list[str], optional): List the type of ops that will be 
+                quantized. Default is ["elementwise_add", "pool2d", "concat"]. 
+            is_full_quantized(bool, optional): If set is_full_quantized be True, apply
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: If set is_full_quantized as True

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 26]
```diff
@@ -1183,77 +1221,111 @@ def _scale_name(self, var_name):
...
 
 class AddQuantDequantPass(object):
+    _supported_quantizable_op_type = [
+        "pool2d", "elementwise_add", "concat", "softmax", "argmax", "transpose",
+        "equal", "gather", "greater_equal", "greater_than", "less_equal",
+        "less_than", "mean", "not_equal", "reshape", "reshape2",
+        "bilinear_interp", "nearest_interp", "trilinear_interp", "slice",
+        "squeeze", "elementwise_sub"
+    ]
+
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py:1242]:
```python
# This is needed by the server GPU int8 inference
_activation_type = ["relu", "relu6", "leaky_relu", "tanh", "swish"]
```

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py [hunk 27]
```diff
@@ -1183,77 +1221,111 @@ def _scale_name(self, var_name):
...
-        self._quantizable_op_type = quantizable_op_type
+
+        if is_full_quantized:
+            self._quantizable_op_type = \
+                AddQuantDequantPass._supported_quantizable_op_type
+        else:
+            self._quantizable_op_type = quantizable_op_type
+            for op_type in quantizable_op_type:
+                assert op_type in AddQuantDequantPass._supported_quantizable_op_type, \
+                    op_type + " is not supported for quantization."
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/quantization_pass.py:1289]:
```python
                (assert op_type in (AddQuantDequantPass._supported_quantizable_op_type + AddQuantDequantPass._activation_type),
                    op_type + " is not supported for quantization.")
```

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 20 above)
Comment (wzzju) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py]: Not done.请改回：
```python
ops = graph.all_op_nodes()
for op_node in ops:
```
这样逻辑比较清晰，表明是对原op节点进行的遍历。

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py [hunk 28]
```diff
@@ -98,21 +104,26 @@ def __init__(self,
...
+        supported_quantizable_op_type = \
+            QuantizationTransformPass._supported_quantizable_op_type + \
+            AddQuantDequantPass._supported_quantizable_op_type
+        if is_full_quantize:
+            self._quantizable_op_type = supported_quantizable_op_type
+        else:
+            self._quantizable_op_type = quantizable_op_type
+            for op_type in self._quantizable_op_type:
+                assert op_type in supported_quantizable_op_type, \
+                    op_type + " is not supported for quantization."
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py:119]:
```python
                assert op_type in (supported_quantizable_op_type + AddQuantDequantPass._activation_type), \
                    op_type + " is not supported for quantization."
```

**Review by wzzju — CHANGES_REQUESTED** [review]

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py (hunk 28 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py:119]: > ```python
>                 assert op_type in (supported_quantizable_op_type + AddQuantDequantPass._activation_type), \
>                     op_type + " is not supported for quantization."
> ```

Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 12 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py:630]: > Not done. 添加QuantizationFreezePass的类变量`_supported_quantizable_op_type = QuantizationTransformPass._supported_quantizable_op_type`。

Haha, Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 26 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py:1242]: > ```python
> # This is needed by the server GPU int8 inference
> _activation_type = ["relu", "relu6", "leaky_relu", "tanh", "swish"]
> ```

Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_pass.py (hunk 27 above)
Comment (juncaipeng) [reply python/paddle/fluid/contrib/slim/quantization/quantization_pass.py:1289]: > ```python
>                 (assert op_type in (AddQuantDequantPass._supported_quantizable_op_type + AddQuantDequantPass._activation_type),
>                     op_type + " is not supported for quantization.")
> ```

Done

**Review by juncaipeng — COMMENTED** [review]
All Done

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py [hunk 29]
```diff
@@ -199,29 +215,30 @@ def _prepare(self):
...
-                    x_var_name = op.input("X")[0]
-                    y_var_name = op.input("Y")[0]
-                    if x_var_name not in persistable_var_names and \
-                        y_var_name not in persistable_var_names:
+                    if self._is_input_all_not_persistable(
+                            op, persistable_var_names):
                         op._set_attr("skip_quant", True)
-                        _logger.warning("A mul op skip quant for two "
+                        _logger.warning("Skip quant a mul op for two "
                                         "input variables are not persistable")
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py:222]: 这里想表达的是“跳过两个输入均非persistable的mul“，这句话好好想一下，应该怎么表达？例如：
```
Detect the mul op has two non-persistable inputs and skip it.
```

File: python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py [hunk 30]
```diff
@@ -319,10 +335,25 @@ def _update_program(self):
...
+    def _save_output_scale(self):
+        '''
+        Save output scale to the quantized op.
+        '''
+        output_scale_name = "output_scale"
+        for op in self._program.global_block().ops:
+            if op.type in self._quantizable_op_type:
+                output_name_list = self._op_real_in_out_name[op.type][1]
+                for output_name in output_name_list:
+                    output_var_name = op.output(output_name)[0]
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/post_training_quantization.py:351]:
```python
for output_var_name in op.output(output_name):
```

**Review by wzzju — APPROVED** [review]
Mostly LGTM.

**Pull request merged**